File: Java/VectorCalculator.java
public class VectorCalculator {
      /**
       * @return[0] = x, @return[1] = y, @return[2] = z
       */
      public static double[] getForwardVector(double pitch, double yaw) {
            return new double[] {
                  (Math.cos(pitch) * Math.sin(yaw)),
                  (Math.sin(pitch) * -1),
                  (Math.cos(pitch) * Math.cos(yaw))
            };
      }

      public static double[] getRightVector(double yaw) {
            return new double[] {
                  Math.cos(yaw),
                  (double)0.0,
                  Math.sin(yaw) * -1
            };
      }

      public static double[] getUpVector(double pitch, double yaw) {
            return cross(
                  getForwardVector(pitch, yaw),
                  getRightVector(yaw)
            );
      }


      public static float[] getForwardVectorF(float pitch, float yaw) {
           return new float[] {
               Math.cos(pitch) * Math.sin(yaw),
               Math.sin(pitch) * -1,
               Math.cos(pitch) * Math.cos(yaw)
           }
      }

      public static float[] getRightVectorF(float yaw) {
           return new double[] {
               Math.cos(yaw),
               (float)0.0,// forgot what 0.0 defaults to
               Math.sin(yaw) * -1
           }
      }

      public static float[] getUpVectorF(float pitch, float yaw) {
           return crossF(
                getForwardVectorF(pitch, yaw),
                getRightVectorF(yaw)
           );
      }

      private static double[] cross(double[] vector1, double[] vector2) {
            double[] r = new double[3];// emphasis on "know your usecases" lol
            for (int i = 0; i < 3; i++) {
                  r[i] = vector1[i] * vector2[i];
            }
            return r;
      }

      private static float[] crossF(float[] vector1, float[] vector2) {
            float[] r = new float[3];// emphasis on "know your usecases" lol
            for (int i = 0; i < 3; i++) {
                  r[i] = vector1[i] * vector2[i];
            }
            return r;
      }
}

File: C/vectorCalculator.c
#include <math.h>
#include <stdlib.h>
#include "vectorCalculator.h"

// NOTE: Returned arrays are malloced, call free() when done using them.

float* getForwardVector(double pitch, double yaw) {
      float* r = (float*)(malloc(3 * sizeof(float)));
      r[0] = (float)(cos(pitch) * sin(yaw));
      r[1] = (float)(sin(pitch) * -1);
      r[2] = (float)(cos(pitch) * cos(yaw));
      return r;
}

float* getRightVector(double yaw) {
      float* r = (float*)(malloc(3 * sizeof(float)));
      r[0] = (float)(cos(yaw));
      r[1] = 0.0;
      r[2] = (float)(sin(yaw) * -1);
      return r;
}

float* getUpVector(double pitch, double yaw) {
      float* forward = getForwardVector(pitch, yaw);
      float* right = getRightVector(yaw);
      float* r = (float*)(malloc(3 * sizeof(float)));
      for (unsigned int i = 0; i < 3; i++) {
            r[i] = forward[i] * right[i];
      }
      free(forward);
      free(right);
      return r;
}
